.github/actions/update-extension-provider-images/update-images.py

```diff
@@ -0,0 +1,691 @@
+
+import semver
+import yaml
+
+from oci import client as oci_client
+from version import parse_to_semver, is_final, iter_upgrade_path
+
+
+@dataclass
+class Update:
+    """A structured representation of a single image update."""
+
+    image_name: str
+    new_tag: str
+    update_type: str  # 'patch', 'minor', or 'singleton'
+    old_tag: Optional[str] = None
+
+
+# --- Helper Functions ---
+def validate_images_data(images_data: Dict[str, Any]) -> None:
+    """
+    Validate the structure of images.yaml data.
+
+    Args:
+        images_data: The loaded YAML data to validate
+
+    Raises:
+        ValueError: If the data structure is invalid or missing required fields
+    """
+    if not isinstance(images_data, dict):
+        raise ValueError("Images data must be a dictionary")
+
+    if "images" not in images_data:
+        raise ValueError("Images data must contain 'images' key")
+
+    if not isinstance(images_data["images"], list):
+        raise ValueError("'images' must be a list")
+
+    required_fields = ["name", "repository", "tag"]
+    for i, image in enumerate(images_data["images"]):
+        for field in required_fields:
+            if field not in image:
+                raise ValueError(f"Image at index {i} missing required field: {field}")
+
+
+# --- Core Logic Functions ---
+def find_greater_versions(
+    current_tags: List[str],
+    available_tags: List[str],
+) -> Dict[str, List[str]]:
+    """
+    Compare current tags with available tags to find greater patch and minor/major versions.
+
+    This function analyzes semantic versions to categorize available updates:
+    - Patch updates: Same major.minor, higher patch version
+    - Minor/Major updates: Higher major or minor version
+
+    Args:
+        current_tags: A list of current tag strings from images.yaml
+        available_tags: A list of all available tag strings from the repository
+
+    Returns:
+        A dictionary with 'patch' and 'minor' keys containing lists of new tags as strings
+    """
+    available_versions_map = {}
+    for tag in available_tags:
+        try:
+            ver = parse_to_semver(tag)
+            if is_final(ver):
+                available_versions_map[ver] = tag
+        except ValueError:
+            continue
+
+    available_versions = sorted(available_versions_map.keys())
+
+    current_versions = []
+    for tag in current_tags:
+        try:
+            current_versions.append(parse_to_semver(tag))
+        except ValueError:
+            continue
+    current_versions.sort()
+
+    if not current_versions:
+        existing_minor_tags = []
+        for ver in available_versions:
+            tag = available_versions_map[ver]
+            existing_minor_tags.append(tag)
+        return {
+            "patch": [],
+            "minor": existing_minor_tags,
+        }
+
+    highest_current_ver = current_versions[-1]
+
+    highest_patch_for_minor = defaultdict(lambda: semver.Version(0))
```

Comment:
I think that's slightly hacky.... how about using `version.find_greatest_version_with_matching_minor` instead? I think this should do the same